### PR TITLE
Add JsonIncludeAttribute & support for non-public accessors

### LINKED ISSUE #29743: JsonSerializer should support private setters as an opt-in feature
Updated by @layomia:
*There's a lot of overlap between this feature and [field support](https://github.com/dotnet/runtime/issues/876) wrt. public API proposal. See https://github.com/dotnet/runtime/issues/34558 for combined API proposal and discussion.*

---

Non-public setters for public properties are not used when deserializing. Other serializers have this feature. As a workaround, user's have to write custom converter's for such POCOs, which isn't trivial in case of a complex object. This feature provides an opt-in for the serializer to use non-public setters.

Enabling non-public getter usage is included for parity with `Newtonsoft.Json` which supports this when the `JsonProperty` attribute is used, and to prevent complicating the API surface if this is ever desired in the future.

This feature was scoped out of v1 (3.x), as the objective was to support simple POCOs. We elected to make this feature opt-in due to security concerns with non-public support.

This feature does not include support for non-public properties.

A related feature that allows deserialization of immutable objects through parameterized constructors was recently added in https://github.com/dotnet/runtime/pull/33444.

### LINKED ISSUE #34453: CreatePropertyInfoForClassInfo run twice?
CreatePropertyInfoForClassInfo run twice?

See 90 and 97, arguments are equivalent:
https://github.com/dotnet/runtime/blob/4a804959512a76556add344d998f3fc9ca9513e8/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs#L90-L97

### PR DESCRIPTION
**Pull request by layomia**
Fixes https://github.com/dotnet/runtime/issues/29743.
Fixes https://github.com/dotnet/runtime/issues/34453 (since I'm in this code).

New API approved in https://github.com/dotnet/runtime/issues/34558.

### CODE REVIEW

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs [hunk 1]
```diff
@@ -382,5 +361,51 @@ public bool DetermineExtensionDataProperty(Dictionary<string, JsonPropertyInfo>
 
             return converter;
         }
+
+        public bool DetermineExtensionDataProperty(Dictionary<string, JsonPropertyInfo> cache)
```
Comment (steveharter) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs:365]: Why were these methods moved (alphabetical sorting?)

Were there any changes (I didn't see any)?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs [hunk 2]
```diff
@@ -189,6 +130,90 @@ public JsonClassInfo(Type type, JsonSerializerOptions options)
...
+                cacheArray = new JsonPropertyInfo[cache.Count];
+            }
+
+            // Set fields when finished to avoid concurrency issues.
+            PropertyCache = cache;
+            cache.Values.CopyTo(cacheArray, 0);
+            PropertyCacheArray = cacheArray;
+        }
+
+        private void HandleNonPublicProperties()
```
Comment (steveharter) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs:198]: This should be able to be combined with the HandlePublicProperties

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs (hunk 1 above)
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs:365]: Arranged the methods mostly in the order that they are used from the constructor, but will revert as there's no real need to do this.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs (hunk 2 above)
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs:198]: Do you mean can combine the logic into one method...

or do you mean we can include the `BindingFlags.NonPublic` flag in the `GetProperties` call and do some if-check to see if the PropertyInfo is non-public? It doesn't seem like there's any property on `PropertyInfo` that can tell us if the property is non-public...

Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs]: (Pushed a commit to minimize the diff in the ctor by inlining the non-public property/`JsonInclude` validation logic).

Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs]: Looks like we can use `IsPublic` on the getters/setters:

```csharp
private static bool IsNonPublicProperty(PropertyInfo propertyInfo)
{
    MethodInfo? getMethod = propertyInfo.GetMethod;
    MethodInfo? setMethod = propertyInfo.SetMethod;
    return !((getMethod != null && getMethod.IsPublic) || (setMethod != null && setMethod.IsPublic));
}
```

Will push a commit

File: src/libraries/System.Text.Json/tests/Serialization/PropertyVisibilityTests.NonPublicAccessors.cs [hunk 3]
```diff
@@ -0,0 +1,332 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using Xunit;
+
+namespace System.Text.Json.Serialization.Tests
+{
+    public static partial class PropertyVisibilityTests
```
Comment (steveharter) [inline src/libraries/System.Text.Json/tests/Serialization/PropertyVisibilityTests.NonPublicAccessors.cs:10]: Currently this PR assumes that the property is still `public`, correct?(meaning the setter or getter can be non-public, but not both). Do we have tests with a private property where both setter and getter are private to verify?

Comment (layomia) [reply src/libraries/System.Text.Json/tests/Serialization/PropertyVisibilityTests.NonPublicAccessors.cs:10]: Yes, I added some tests showing that non-public properties (with various accessor visibility levels) are not included: https://github.com/dotnet/runtime/blob/e4e05858e9689099b907387a1159de6a997d65d7/src/libraries/System.Text.Json/tests/Serialization/PropertyVisibilityTests.cs#L869-L891

These tests ensure that we throw `IOE` when `[JsonInclude]` is used on a non-public properties: https://github.com/dotnet/runtime/blob/e4e05858e9689099b907387a1159de6a997d65d7/src/libraries/System.Text.Json/tests/Serialization/PropertyVisibilityTests.NonPublicAccessors.cs#L295-L330

**Review by steveharter — APPROVED** [review]

**Comment by TanvirArjel** [comment]
@layomia Is this available on .NET 5.0 RC1, maybe not. My Project has been updated to .NET 5.0 RC1 but it's not working yet.

**Comment by layomia** [comment]
@TanvirArjel the feature is in RC1, but it requires explicitly opting in with `JsonIncludeAttribute`:

```cs
public class DataTableParamsDto
{
    public int Draw { get; set; }

    public int Start { get; set; }

    public int Length { get; set; }

    [JsonInclude]
    public List<ColumnRequestItem> Columns { get; private set; }

    [JsonInclude]
    public List<OrderRequestItem> Order { get; private set; }

    [JsonInclude]
    public SearchRequestItem Search { get; private set; }
}
```

**Pull request merged**